### PR TITLE
Webserver

### LINKED ISSUE #8: Host some sort of webserver to interact/display information from the bot

### LINKED ISSUE #29: Store name of the user who added a song
**Is your feature request related to a problem? Please describe.**
If a user joins the channel after the playlist has been filled they can't see who requested a song.

**Describe the solution you'd like**
On add the bot should check what the name of the message sender was and store it to display it later. This could be used in multiple places like in the message that gets sent when a new song starts.

### PR DESCRIPTION
**Pull request by Jokler**
**Summary**
This PR implements the following **features**

- Webserver based on actix
- Adds duration, thumbnail and the name of the user who requested a url to `AudioMetadata`
- Easily extendable with more front-ends the user can switch between
- Demo site to show of the information available
- Tmtu-mode, a simple site based on tables (without a button to leave it again) created by @fkaa 
- Simple API to display the available information
- Replaces all Mutexes with RwLocks to allow the website to read from multiple threads at once

**Closing issues**
Resolves #8
Resolves #29

### CODE REVIEW

File: src/playlist.rs [hunk 1]
```diff
@@ -28,6 +28,16 @@ impl Playlist {
         res
     }
 
+    pub fn to_vec(&self) -> Vec<AudioMetadata> {
+        let (a, b) = self.data.as_slices();
+
+        let mut res = a.to_vec();
+        res.extend_from_slice(b);
+        res.reverse();
```
Comment (Jokler) [inline src/playlist.rs:36]: There is probably a better way but I can't be bothered right now.

File: Cargo.toml [hunk 2]
```diff
@@ -27,11 +27,17 @@ futures-preview = { version = "=0.3.0-alpha.19", features = ["compat"] }
...
-rand = { version = "0.7.3", features = ["small_rng"] }
+serde_json      = "1.0.44"
+serde           = "1.0.104"
+actix           = "0.9.0"
+actix-rt        = "1.0.0"
+actix-web       = "2.0.0"
+actix-files     = "0.2.1"
+derive_more     = "0.99.2"
+askama          = { version = "0.9.0", features = ["with-actix-web"] }
+rand            = { version = "0.7.3", features = ["small_rng"] }
```
Comment (fkaa) [inline Cargo.toml]: no indenting please

File: src/audio_player.rs [hunk 3]
```diff
@@ -104,15 +108,21 @@ impl AudioPlayer {
 
         pipeline.add(&audio_bin)?;
 
+        // The documentation says that we have to make sure to handle
+        // all messages if auto flushing is deactivated.
+        // I hope our way of reading messages is good enough.
+        pipeline.set_auto_flush_bus(false);
```
Comment (fkaa) [inline src/audio_player.rs]: link?

File: src/web_server.rs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+//    pub fn default(
+//        opt: &Option<impl std::fmt::Display>,
+//        def: &dyn std::fmt::Display,
+//    ) -> Result<String, askama::Error> {
+//        match opt {
+//            Some(d) => Ok(d.to_string()),
+//            None => Ok(def.to_string()),
+//        }
+//    }
```
Comment (fkaa) [inline src/web_server.rs]: unused?

**Review by fkaa — CHANGES_REQUESTED** [review]
Run cargo fmt and put web stuff in its own folder instead of the root

File: src/audio_player.rs (hunk 3 above)
Comment (Jokler) [reply src/audio_player.rs]: added

**Pull request merged**